Expressions: factor out the lexer

nuoHep opened the pull request:
Refactor expression tokenization by factoring out a lexer to make it
easier to follow.

The lexer is non-allocating - tokens contain views into the input
string view.

Drop an intermidate `convert_to_unary()` step as the lexer produces
unary tokens for plus/minus, so we need to convert to binary, and it's
actually easier to do that right inside `parse_tokens()`. We never use
non-converted token streams anyway.

---

Plus a smol cleanup of token printing on top.

nuoHep committed:
Expressions: factor out the lexer

Refactor expression tokenization by factoring out a lexer to make it
easier to follow.

The lexer is non-allocating - tokens contain views into the input
string view.

Drop an intermidate `convert_to_unary()` step as the lexer produces
unary tokens for plus/minus, so we need to convert to binary, and it's
actually easier to do that right inside `parse_tokens()`. We never use
non-converted token streams anyway.

nuoHep committed:
Expressions: refactor token printing

scamille commented on engine/sim/sc_expressions.cpp:
```diff
@@ -16,6 +16,145 @@ namespace
+  bool match( util::string_view str, int consume_mod = 0 )
+  {
+    if ( util::str_prefix_ci( input.substr( current_len ), str ) ) {
+      current_len += ( str.size() + consume_mod );
```
Maybe add an assert here that the resulting calculation is still in bounds of the input.
I understand that for the current match calls with consume_mod = -1 this should always be fine.

scamille commented on engine/sim/sc_expressions.cpp:
```diff
@@ -16,6 +16,145 @@ namespace
+      case '*': return yield_token( TOK_MULT );
+      case '%':
+      {
+        if ( match( '%' ) )
```
So %% is mod, and a single % is division? Might be worth a comment since this is quite a bit unintuitive at first glance.

nuoHep replied:
Yeah, been like that since d68425e69c0c032053b7cafe43cfa8727596c78c
Will add (I think it's kinda self documenting, but I see how it can be non-obvious when looking at it for the first time).

nuoHep replied:
Added a comment for why we use `%`/`%%` for div/mod in b941c44

scamille commented on engine/sim/sc_expressions.cpp:
```diff
@@ -1035,69 +1011,65 @@ bool is_binary( token_e expr_token_type )
-  std::string str;
-  if ( !tokens.empty() )
-    str += "tokens: ";
+    // optimization: fold minus followed by a number together into a single number token
```
Couldn't this be done inside the lexer?
Maybe move the TOK_NUM matching (line 75 to 83) to a helper function, and then also call that when you have a TOK_MINUS match.

nuoHep replied:
I'd rather not tbh.
`-` can be both unary (`TOK_MINUS`) and binary (`TOK_SUB`). Right now the lexer does not care and always produces only the unary one. Moving this to the lexer will necessarily require the lexer to also handle binary `-` (as `)-1` should produce `[TOK_RPAR, TOK_SUB, TOK_NUM]`) which would require adding lookbehind.

I actually thought about handling `ceil` & `floor` outside of the lexer too (they can be easily matched outside), don't remember why I decided to leave it here (the commit is from July when I was playing with an interpreter for expressions).

scamille replied:
Oh ok. So that's the main reason to restrict the lexer to unary is so that it works by just looking forward. And then it is easier to build binary SUB/ADD after the lexing pass and then optimize the leftover unary Minus+Num tokens.

Makes a lot of sense now that I fully understand it.

nuoHep replied:
Right, outside of "linguistic purity" reasons, this is mostly about making the lexer simpler.

It's mostly a trade-off tbh. You still need to handle the "duality" of `-` (the fact that it's unary/binary depending on context) somewhere. It just so happens that handling it outside of the lexer is simpler - you have to handle the duality of `+` too and parsing is typically structured in such a way there is implicit "lookbehind".

If we used a more sophisticated parser this would be pretty obvious.

scamille replied:
Thanks again for all the explanations.

scamille commented:
Great rewrite. Makes it much easier to understand the entirety of the simc lexing.

nuoHep committed:
Expressions: comment why we use '%'/'%%' for div/mod

nuoHep committed:
Expressions: move FLOOR/CEIL handling out of the lexer

FLOOR/CEIL are a bit special in that they really aren't "tokens"
(lexemes) in the classical sense. The reason we do them as tokens seems
mostly to be because it was just easier to do them this way and to make
expressions (which are basically AST nodes) have a unique token type for
them.

In reality they both are "pseudo-functions" - a special case of a
generic single argument function - a [TOK_STR, TOK_LPAR] sequence with a
"known" value for TOK_STR (which is actually an "identifier"
(imaginary TOK_IDENT), really).

So handling them in the lexer does not really make sense. It would be
much more apparent if we did single-pass parsing (instead of ~3 passes
we do now).
